modules/core/src/main/java/org/apache/ignite/internal/util/GridConcurrentMultiPairQueue.java

```diff
@@ -0,0 +1,229 @@
+        new GridConcurrentMultiPairQueue<>(Collections.emptyMap());
+
+    /** Inner holder. */
+    private final V[][] vals;
+
+    /** Storage for every array length. */
+    private final int[] lenSeq;
+
+    /** Current absolute position. */
+    private final AtomicInteger pos = new AtomicInteger();
+
+    /** Precalculated max position. */
+    private final int maxPos;
+
+    /** Keys array. */
+    private final K[] keysArr;
+
+    /** Thread local cached segment. */
+    private ThreadLocal<Integer> cachedSegment = ThreadLocal.withInitial(() -> 0);
+
+    /** */
+    public GridConcurrentMultiPairQueue(Map<K, ? extends Collection<V>> items) {
+        int pairCnt = (int)items.entrySet().stream().map(Map.Entry::getValue).filter(k -> k.size() > 0).count();
+
+        vals = (V[][])new Object[pairCnt][];
+
+        keysArr = (K[])new Object[pairCnt];
+
+        lenSeq = new int[pairCnt];
+
+        int keyPos = 0;
+
+        int size = -1;
+
+        for (Map.Entry<K, ? extends Collection<V>> p : items.entrySet()) {
+            if (p.getValue().isEmpty())
+                continue;
+
+            keysArr[keyPos] = p.getKey();
+
+            lenSeq[keyPos] = size += p.getValue().size();
+
+            vals[keyPos++] = (V[])p.getValue().toArray();
+        }
+
+        maxPos = size + 1;
+    }
+
+    /** */
+    public GridConcurrentMultiPairQueue(Collection<T2<K, V[]>> items) {
+        int pairCnt = (int)items.stream().map(Map.Entry::getValue).filter(k -> k.length > 0).count();
+
+        vals = (V[][])new Object[pairCnt][];
+
+        keysArr = (K[])new Object[pairCnt];
+
+        lenSeq = new int[pairCnt];
+
+        int keyPos = 0;
+
+        int size = -1;
+
+        for (Map.Entry<K, V[]> p : items) {
+            if (p.getValue().length == 0)
+                continue;
+
+            keysArr[keyPos] = p.getKey();
+
+            lenSeq[keyPos] = size += p.getValue().length;
+
+            vals[keyPos++] = p.getValue();
+        }
+
+        maxPos = size + 1;
+    }
+
+    /**
+     * Retrieves and removes the head of this queue,
+     * or returns {@code null} if this queue is empty.
+     *
+     * @return the head of this queue, or {@code null} if this queue is empty
+     */
+    public @Nullable T2<K, V> poll() {
+        int absPos = pos.getAndIncrement();
+
+        if (absPos >= maxPos)
+            return null;
+
+        int segment = cachedSegment.get();
+
+        if (absPos > lenSeq[segment]) {
+            segment = Arrays.binarySearch(lenSeq, segment, lenSeq.length - 1, absPos);;
+
+            segment = segment < 0 ? -segment - 1 : segment;
+
+            cachedSegment.set(segment);
+        }
+
+        int relPos = segment == 0 ? (int)absPos : (int)(absPos - lenSeq[segment - 1] - 1);
+
+        K key = keysArr[segment];
+
+        return new T2<>(key, vals[segment][relPos]);
+    }
+
+    /** Return next initialized input class instance. */
+    public boolean next(Result<K, V> res) {
```

Comment:
I see that this method copy-pasted. Can we reuse code of "poll"? Does someone uses that method? I believe that you only use it in tests.

Reply:
I change signature, now they didn`t clone each other.